src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs

```diff
@@ -160,26 +255,96 @@ private bool TryMatch(IList<ParameterDescriptor> parameters, IList<string> avail
                     }
                 }
 
+                // if we can't find the parameter in the request, check whether
+                // there's a special model binder registered to handle it
+                if (ParameterHasRegisteredModelBinder(p))
+                {
+                    continue;
+                }
+
+                // if parameter is not bound to a key in the path,
+                // assume that it's bound to the request body
+                // only one parameter should be considered bound to the body
+                if (!matchedBody && RequestHasBody(context))
+                {
+                    matchedBody = true;
+                    continue;
+                }
+
                 return false;
             }
 
             return true;
         }
 
+        private bool RequestHasBody(RouteContext context)
+        {
+            string method = context.HttpContext.Request.Method.ToLowerInvariant();
+            return method.Equals("post", StringComparison.OrdinalIgnoreCase) ||
+                method.Equals("put", StringComparison.OrdinalIgnoreCase) ||
+                method.Equals("patch", StringComparison.OrdinalIgnoreCase) ||
+                method.Equals("merge", StringComparison.OrdinalIgnoreCase);
+        }
+
+        private bool ActionAcceptsMethod(ControllerActionDescriptor action, string method)
+        {
+            if (action == null)
```

Comment:
@habbes Shouldn't we instead use a `Debug.Assert()` or `Contract.Assert()` here?